File: java/src/main/java/club/marzipan/javacarrentals/VehicleMapper.java
package club.marzipan.javacarrentals;

import org.springframework.stereotype.Component;

@Component
public class VehicleMapper {
    public Vehicle mapEntityToApi(VehicleEntity vehicleEntity) {
        if (vehicleEntity == null) {
            return null;
        }
        Vehicle vehicle = new Vehicle();
        vehicle.setId(vehicleEntity.getId());
        vehicle.setLicensePlate(vehicleEntity.getLicensePlate());
        vehicle.setVin(vehicleEntity.getVin());
        vehicle.setYear(vehicleEntity.getYear());
        vehicle.setMake(vehicleEntity.getMake());
        vehicle.setVehicleClass(mapEntityToApi(vehicleEntity.getVehicleClass()));
        return vehicle;
    }

    public Vehicle.VehicleClass mapEntityToApi(VehicleEntity.VehicleClass vehicleClass) {
        if (vehicleClass == null) {
            return null;
        }
        switch (vehicleClass) {
            case standard_small:
                return Vehicle.VehicleClass.standard_small;

            case standard_large:
                return Vehicle.VehicleClass.standard_large;

            case premium_small:
                return Vehicle.VehicleClass.premium_small;

            case premium_large:
                return Vehicle.VehicleClass.premium_large;

            default:
                return null;
        }
    }

    public VehicleEntity.VehicleClass mapApiToEntity(Vehicle.VehicleClass vehicleClass) {
        if (vehicleClass == null) {
            return null;
        }
        switch (vehicleClass) {
            case standard_small:
                return VehicleEntity.VehicleClass.standard_small;

            case standard_large:
                return VehicleEntity.VehicleClass.standard_large;

            case premium_small:
                return VehicleEntity.VehicleClass.premium_small;

            case premium_large:
                return VehicleEntity.VehicleClass.premium_large;

            default:
                return null;
        }
    }
}

File: rust/src/vehicle_mapper.rs
use crate::{
    vehicle::{Vehicle, VehicleClass},
    vehicle_entity::{VehicleClass as VehicleClassEntity, VehicleEntity},
};

impl From<VehicleEntity> for Vehicle {
    fn from(vehicle_entity: VehicleEntity) -> Vehicle {
        Vehicle {
            id: vehicle_entity.id.into(),
            license_plate: vehicle_entity.license_plate.into(),
            vin: vehicle_entity.vin.into(),
            year: vehicle_entity.year as _,
            make: vehicle_entity.make.into(),
            model: vehicle_entity.model.into(),
            vehicle_class: vehicle_entity.vehicle_class.into(),
        }
    }
}

impl From<VehicleClassEntity> for VehicleClass {
    fn from(status: VehicleClassEntity) -> VehicleClass {
        match status {
            VehicleClassEntity::StandardSmall => VehicleClass::StandardSmall,
            VehicleClassEntity::StandardLarge => VehicleClass::StandardLarge,
            VehicleClassEntity::PremiumSmall => VehicleClass::PremiumSmall,
            VehicleClassEntity::PremiumLarge => VehicleClass::PremiumLarge,
        }
    }
}
